### PR TITLE
Add MultiChoice widget

### LINKED ISSUE #9532: Please replace MultiSelect widget with something more compact, effective and nicer looking
ORIGINALLY POSTED AS PANEL FEATURE REQUEST AT https://github.com/holoviz/panel/issues/874

#### My Pain

I know of experience that many of my dashboard have multiple multi-selections. 

The MultiSelect widget of panel is not compact so it takes up a lot of space.

Furthermore I find the MultiSelect not very nicely looking. I would like my dashboards to look appealing and fresh.

Futhermore I think navigating and selecting in the MultiSelect widget is slow as soon as you have to start scrolling in the MultiSelect

image

#### Solution

Implement compact, efficient and nicer looking MultiSelect. 

It should work as a dropdown with multiselect.

#### Additional Context

You can get inspiration from Dash, Streamlit and Tableau that all have a much more compact and modern looking widget.

image

image

FYI. Tableau has both a more compact Dropdown and something similar to the MultiSelect.

Here it's used an that's where I have my evaluation from. You can find it in the Gallery at awesome-panel.org.

kickstarter_dashboard

### PR DESCRIPTION
**Pull request by philippjfr**
Adds a `MultiChoice` widget, which is a more compact `MultiSelect` widget based on the [choices.js](https://github.com/jshjohnson/Choices/) library. choices.js is only 19 KB and does not have a jQuery dependency like other libraries like Select2.

Screen Shot 2019-12-27 at 7 21 22 PM

- [x] issues: fixes #9532 
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by philippjfr** [comment]
Suggestions for naming, styling and exposing other options as properties welcome.

**Comment by bryevdv** [comment]
Def don't like `Pillbox`. Maybe `Chooser`?

I personally like the solid-filled and rounder style on [their demo](https://joshuajohnson.co.uk/Choices/) a bit more, but I'd suggest a default blue of "bokeh blue" (default glyph color). 

If there's not a way to add option grouping, search, or remote fetch right now would be good to at lest think about how they might fit in later without breaking changes.  Otherwise some simple but useful properties to expose include:

* maxItemCount
* renderChoiceLimit
* removeItemButton
* placeholderValue

**Comment by philippjfr** [comment]
>Def don't like Pillbox. Maybe Chooser?

Definitely better, but don't love it. Let's keep brainstorming.

>I personally like the solid-filled and rounder style on their demo a bit more

Let's get a tie-breaker, @mattpap, @jbednar, @MarcSkovMadsen.

>If there's not a way to add option grouping

If we come up with a concrete proposal for declaring the groupings I'd be happy to add it, but I think consistency with the MultiSelect API is also valuable.

> search

is enabled by default.

>remote fetch

Can't currently see how to do it since you need to define an async function somehow.

The other properties do seem easy to expose.

**Comment by jbednar** [comment]
I'm very happy to see this functionality, but I don't think I have any useful naming suggestions. I don't like either Pillbox or Chooser; Pillbox doesn't sound like a widget and Chooser doesn't convey that one can choose multiple.  It looks like Streamlit calls this widget Multiselect, while Dash calls it Dropdown(multi=True). I think it makes more sense to be a separately implemented and named widget than just an option on a dropdown, because it behaves quite differently, so I vote against the Dash naming.  Panel and Bokeh already have a Multiselect that looks and works differently, but the fundamental functionality is similar, so at the Panel level it is conceivable that this new widget is just a display option for the existing Panel Multiselect class. The filtering options are quite different, though, so there's also plenty of argument for keeping it separate at the Panel level. Maybe `Selections` or `Choices`?

Bokeh's widgets are generally quite square, so I think I prefer the styling in this PR to the rounder pill-shaped ones from the choices.js demo.  Supporting the suggested properties sounds great!

**Comment by bryevdv** [comment]
I think having a plural for a widget name is extremely weird to m, and I am a pretty hard -1 on either `SelectionS` or `ChoiceS`. I would even prefer `Pillbox` to those. To me, `Chooser` says "thing I can use to make choices" I definitely would not assume its necessarily only one thing either. If we want to be pedantic, then maybe `MultiChooser` I guess. 

cc @birdsarah @canavandl @hyles-lineata for more opinions on visual style. I'm definitely opposed to "dark colored text on a colored background with a dark colored outline" as in the OP, I think solid color/light-white text is much cleaner and the way to go even if no-one else likes the more rounder style.

**Comment by philippjfr** [comment]
How about `MultiChoice` since it's somewhat more similar to `MultiSelect` (which is also not called `MultiSelector`). I'm okay with `MultiChooser` too though.

**Comment by hyles-lineata** [comment]
I second `MultiChoice`!

**Comment by philippjfr** [comment]
Updated styling:

Screen Shot 2019-12-28 at 2 29 46 AM

**Comment by jbednar** [comment]
I'm happy with MultiChoice and the new styling.

File: bokehjs/src/lib/models/widgets/multichoice.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      multiple: true,
+      class: bk_input,
+      name: this.model.name,
+      disabled: this.model.disabled
+    }, options)
+
+    this.group_el.appendChild(this.select_el)
+	this.render_selection()
+
+    let opts: ChoicesOpts = {
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/multichoice.ts]: `const`.

File: bokehjs/src/less/widgets/choices.less [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    border-radius: 5px;
+    min-height: unset;
+  }
+
+  .choices[data-type*=select-multiple] .choices__button {
+    background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjEiIGhlaWdodD0iMjEiIHZpZXdCb3g9IjAgMCAyMSAyMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmZmZmIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPjxwYXRoIGQ9Ik0yLjU5Mi4wNDRsMTguMzY0IDE4LjM2NC0yLjU0OCAyLjU0OEwuMDQ0IDIuNTkyeiIvPjxwYXRoIGQ9Ik0wIDE4LjM2NEwxOC4zNjQgMGwyLjU0OCAyLjU0OEwyLjU0OCAyMC45MTJ6Ii8+PC9nPjwvc3ZnPg==);
+	border-left: 1px solid white;
+	opacity: 1;
+  }
+}
```
Comment (mattpap) [inline bokehjs/src/less/widgets/choices.less]: Please normalize indentation.

File: bokehjs/src/lib/models/widgets/multichoice.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+import {select, option} from "core/dom"
+import {isString} from "core/util/types"
+import * as p from "core/properties"
+import {bk_input} from "styles/widgets/inputs"
+import "styles/widgets/choices"
+
+import {InputWidget, InputWidgetView} from "./input_widget"
+
+export type ChoicesOpts = {
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/multichoice.ts]: This should be imported from `choices.js`.

**Comment by mattpap** [comment]
This component seems huge:
```
   js/bokeh-widgets.min.js file size
      1) should be 174 kB <= 90 kB
```
It may be possible to import its necessary sub-components more selectively to reduce size.

**Comment by philippjfr** [comment]
That seems massive indeed. I can't seem to figure out how import just certain sub-components. Almost doubling the widgets bundle doesn't seem like a good option so maybe I should leave it in Panel for now?

**Comment by bryevdv** [comment]
It's a nice widget and lacking it we lack parity with other competing tools. I would say put it in its own bundle (e.g like tables)

**Comment by bryevdv** [comment]
Also I prefer `Chooser` over `MultiChoice` tho I could live with the latter. I'd prefer to solicit more opiniokns for now, though.

**Comment by birdsarah** [comment]
I don't like Chooser, it's very ambiguous. How is choose different from selector. For me, as @philippjfr  said initially this "is a more compact MultiSelect." For me this feels like a styling / theming / ?? variation on MultiSelect. Not likely to be a popular opinion but I'm just throwing it out there.

On a slightly broader note, this feels like a pretty random addition - adding one implementation from a completely new widget library. I agree with the point that @bryevdv made that "we lack parity with other competing tools." Can we have a more systematic way to wrap a whole library (bootstrap, choices, etc) as a separate installable component and then include it and just borrow the names from those libraries as they're used - does that make sense?

**Comment by bryevdv** [comment]
>  For me this feels like a styling / theming / ?? variation on MultiSelect. Not likely to be a popular opinion but I'm just throwing it out there.

I don't really disagree with this, but it's also not quite identical either and I expect trying to fit both under an identical API would take more work than we can presently muster in the short term.  If @philippjfr want's to combine this under `MultiSelect` I am not opposed, tho I think that would make it harder to ever split into a separate component (see below). 

> Can we have a more systematic way to wrap a whole library (bootstrap, choices, etc) as a separate installable component and then include it and just borrow the names from those libraries as they're used - does that make sense?

The idea makes sense, but I think that would require multiple people to come together for a non-trivial amount of strategic planning, including people that have good/strong opinions about what it should look like. 

FWIW I suggested `Chooser` because the name of underlying library is *Choices.js*. Along these lines what about just calling it `Choices` then?

As for separately installable component, AFAIK @mattpap current work is to make bundles more fine-grained (though now WIP has been shared so I can't say with certainty). In which case I would certainly support putting this component in its own bundle. The issue I see with making it *separately installable* is that to me that means its not published on our CDN alongside our "built in" bundles, and giving up CDN  benefits (esp. no caching) seems like a steep price for users.

**Comment by philippjfr** [comment]
```
  1) bokehjs/build/*/*.min.js file sizes
       js/bokeh-widgets.min.js file size
         should be 241 kB <= 160 kB:
```

How come the bundle has grown so much? This PR already grows it but it already almost doubled in size since I last worked on the PR.

**Comment by mattpap** [comment]
> How come the bundle has grown so much?

New date picker widget is the reason. Just increase the size once more (250 kB should be fine).

**Comment by philippjfr** [comment]
Ah thanks! So one different question, what's the equivalent of selenium tests now, should I add some tests like that and what's the best example of those to look at?

**Comment by mattpap** [comment]
> what's the equivalent of selenium tests now, should I add some tests like that and what's the best example of those to look at?

Currently the best bet is on `tests/integration`. There are also integration tests in `bokehjs/test/integration`, but only "static" ones at this point (that test framework is WIP). Also, bokehjs' unit tests (`bokehjs/test/unit`) now run in a browser (as opposed to nodejs previously), so they also can act pretty much as integration tests. In fact both use the same infrastructure, so the difference is more organizational in nature than anything. If you want to just generate a pretty image under image diff tests, then you can add an integration example under `examples/integration`. Note that currently image diff is broken after CI changes. Those integration examples will be merged into bokehjs' integration tests as soon as that framework supports image diff (currently it only supports test asserts and baseline diffs).

In general my goal is to move all functional testing of bokehjs' features to bokehjs at some point (as soon as testing framework in bokehjs is capable and stable enough). In Python we will test only the APIs and high-level integration with bokehjs (aka. if serialization and protocol work).

**Comment by philippjfr** [comment]
@bryevdv Added some tests, let me know if you think we'll need more but I need to switch to some other stuff now.

**Comment by bryevdv** [comment]
@philippjfr BokehJS size check failures are real

**Pull request merged**